### PR TITLE
Markup by country of location of the hotel for a specific supplier

### PR DESCRIPTION
**Pull request by DaniyarAskarov**
Issue: https://github.com/happy-travel/agent-app-project/issues/1376
Insomnia: Edo -> Admin -> Markups -> Supplier Markups
Tested through unit tests

Test case:

1) Add supplier destination country markup through POST endpoint ../admin/supplier-markups and check validators
2) Start availability search for hotel from country for specific supplier which markup was added before
3) Check if markup was applied
JSON example:
{
"description": "test",
"value": 5,
"destinationScopeId": "RU",
"destinationScopeType": 2,
"supplierCode": "jumeirah"
}

### CODE REVIEW

File: Api/Services/Markups/MarkupPolicyService.cs [hunk 1]
```diff
@@ -33,29 +34,60 @@ public List<MarkupPolicy> Get(MarkupSubjectInfo subjectInfo, MarkupDestinationIn
...
+
+
+            static bool IsApplicableBySupplier(MarkupPolicy policy, MarkupDestinationInfo info)
+                => policy.SupplierCode == info.SupplierCode;
+        }
+
+
+        private static bool IsApplicableBySubject(MarkupPolicy policy, MarkupSubjectInfo info) => policy.SubjectScopeType switch
+        {
+            // This code will be uncommented at the second stage of work on markups - Issue - AA #1310
```
Comment (AlexeyShok) [inline Api/Services/Markups/MarkupPolicyService.cs:70]: Extra comment

**Review by AlexeyShok — APPROVED** [review]

File: Api/Services/Markups/IMarkupPolicyService.cs [hunk 2]
```diff
@@ -9,5 +9,6 @@ namespace HappyTravel.Edo.Api.Services.Markups
     public interface IMarkupPolicyService
     {
         List<MarkupPolicy> Get(MarkupSubjectInfo subjectInfo, MarkupDestinationInfo destinationInfo);
+        List<MarkupPolicy> GetSecondLevel(MarkupSubjectInfo subjectInfo, MarkupDestinationInfo destinationInfo);
```
Comment (kirillta) [inline Api/Services/Markups/IMarkupPolicyService.cs:12]: What second level means?

File: Api/Services/Markups/MarkupPolicyService.cs [hunk 3]
```diff
@@ -23,7 +22,9 @@ public List<MarkupPolicy> Get(MarkupSubjectInfo subjectInfo, MarkupDestinationIn
             agencyTreeIds.Add(agencyId);
 
             var policies = _markupPolicyStorage.Get(policy =>
-                IsApplicableBySubject(policy, subjectInfo) && IsApplicableByObject(policy, destinationInfo)
+                IsApplicableBySubject(policy, subjectInfo) &&
+                IsApplicableByObject(policy, destinationInfo) &&
```
Comment (kirillta) [inline Api/Services/Markups/MarkupPolicyService.cs:26]: `Subject` and `Object` are unclear

File: Api/Services/Markups/MarkupService.cs [hunk 4]
```diff
@@ -28,31 +29,42 @@ public class MarkupService : IMarkupService
...
+                // Second level markups appliement
+                value = value * (100 + secondLevelPercentSum) / 100;
                 return new ValueTask<MoneyAmount>(new MoneyAmount(value, initialPrice.Currency));
             };
 
 
             return await priceProcessFunc(details, percentMarkupFunction);
 
+
+            decimal GetLevelsPercentSum(Func<MarkupSubjectInfo, MarkupDestinationInfo, List<MarkupPolicy>> policyGetFunc)
```
Comment (kirillta) [inline Api/Services/Markups/MarkupService.cs:48]: Levels again

Comment (kirillta) [reply Api/Services/Markups/MarkupService.cs:48]: `getPolicyFunc`

File: Api/Services/Markups/MarkupService.cs [hunk 5]
```diff
@@ -28,31 +29,42 @@ public class MarkupService : IMarkupService
...
+            {
+                var policies = policyGetFunc(subject, destinationInfo);
+
+                var percentPolicies = policies
+                    .Where(p => p.FunctionType == MarkupFunctionType.Percent)
+                    .ToList();
+
+                var percentSum = percentPolicies.Sum(p => p.Value);
```
Comment (kirillta) [inline Api/Services/Markups/MarkupService.cs:56]: Could this lime be attached to one above?

File: Api/Services/Markups/MarkupService.cs [hunk 6]
```diff
@@ -28,31 +29,42 @@ public class MarkupService : IMarkupService
...
+                if (percentSum <= 0)
+                {
+                    var policiesSlim = policies.Select(p => new { Id = p.Id, Value = p.Value }).ToList();
```
Comment (kirillta) [inline Api/Services/Markups/MarkupService.cs:59]: `slimPolicies`. Please, use more natural language

File: Api/Services/Markups/MarkupPolicyService.cs [hunk 7]
```diff
@@ -33,29 +34,60 @@ public List<MarkupPolicy> Get(MarkupSubjectInfo subjectInfo, MarkupDestinationIn
...
+            static bool IsApplicableBySupplier(MarkupPolicy policy, MarkupDestinationInfo info)
+                => policy.SupplierCode == info.SupplierCode;
+        }
+
+
+        private static bool IsApplicableBySubject(MarkupPolicy policy, MarkupSubjectInfo info) => policy.SubjectScopeType switch
+        {
+            // This code will be uncommented at the second stage of work on markups - Issue - AA #1310
+            SubjectMarkupScopeTypes.Global => true,
+            SubjectMarkupScopeTypes.Market => policy.SubjectScopeId == info.MarketId.ToString(),
```
Comment (kirillta) [inline Api/Services/Markups/MarkupPolicyService.cs:72]: Why cast?

**Review by kirillta — APPROVED** [review]

File: Api/Controllers/AdministratorControllers/SupplierMarkupController.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+        [AdministratorPermissions(AdministratorPermissions.MarkupManagement)]
+        public async Task<IActionResult> AddPolicy([FromBody] SupplierMarkupRequest request)
+        {
+            var (_, isFailure, error) = await _policyManager.Add(request);
+            if (isFailure)
+                return BadRequest(ProblemDetailsBuilder.Build(error));
+
+            return NoContent();
```
Comment (AlexeyShok) [inline Api/Controllers/AdministratorControllers/SupplierMarkupController.cs:43]: It's better to use NoContentOrBadRequest extension

File: Api/Controllers/AdministratorControllers/SupplierMarkupController.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        /// <summary>
+        ///     Gets supplier markup policies
+        /// </summary>
+        /// <returns>List of supplier markups</returns>
+        [HttpGet]
+        [ProducesResponseType(typeof(List<MarkupInfo>), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+        [AdministratorPermissions(AdministratorPermissions.MarkupManagement)]
+        public async Task<IActionResult> GetPolicies()
+            => Ok(await _policyManager.Get());
```
Comment (AlexeyShok) [inline Api/Controllers/AdministratorControllers/SupplierMarkupController.cs:56]: OkOrBadRequest

File: Api/Controllers/AdministratorControllers/SupplierMarkupController.cs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+        [AdministratorPermissions(AdministratorPermissions.MarkupManagement)]
+        public async Task<IActionResult> ModifyPolicy(int policyId, [FromBody] SupplierMarkupRequest request)
+        {
+            var (_, isFailure, error) = await _policyManager.Modify(policyId, request);
+            if (isFailure)
+                return BadRequest(ProblemDetailsBuilder.Build(error));
+
+            return NoContent();
```
Comment (AlexeyShok) [inline Api/Controllers/AdministratorControllers/SupplierMarkupController.cs:75]: It's better to use NoContentOrBadRequest extension

File: Api/Controllers/AdministratorControllers/SupplierMarkupController.cs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+        [AdministratorPermissions(AdministratorPermissions.MarkupManagement)]
+        public async Task<IActionResult> RemovePolicy([FromRoute] int policyId)
+        {
+            var (_, isFailure, error) = await _policyManager.Remove(policyId);
+            if (isFailure)
+                return BadRequest(ProblemDetailsBuilder.Build(error));
+
+            return NoContent();
```
Comment (AlexeyShok) [inline Api/Controllers/AdministratorControllers/SupplierMarkupController.cs:94]: It's better to use NoContentOrBadRequest extension

File: Api/Controllers/AdministratorControllers/SupplierMarkupController.cs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+
+        /// <summary>
+        ///     Gets supplier markup policies
+        /// </summary>
+        /// <returns>List of supplier markups</returns>
+        [HttpGet]
+        [ProducesResponseType(typeof(List<MarkupInfo>), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
```
Comment (AlexeyShok) [inline Api/Controllers/AdministratorControllers/SupplierMarkupController.cs:53]: No 400 return code here

File: Api/Models/Markups/Supplier/SupplierMarkupRequest.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        /// </summary>
+        public string DestinationScopeId { get; init; }
+
+
+        /// <summary>
+        ///     Destination type of booking from the mapper
+        /// </summary>
+        public DestinationMarkupScopeTypes DestinationScopeType { get; init; }
+
+
```
Comment (kirillta) [inline Api/Models/Markups/Supplier/SupplierMarkupRequest.cs:44]: Formatting, please

**Review by AlexeyShok — APPROVED** [review]

File: Api/Services/Markups/SupplierMarkupPolicyManager.cs [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+            => _context.MarkupPolicies
+                .Where(p => !string.IsNullOrEmpty(p.SupplierCode))
+                .Select(p => new MarkupInfo(p.Id, p.GetSettings()))
+                .ToListAsync(cancellationToken);
+
+
+        public async Task<Result> Add(SupplierMarkupRequest request, CancellationToken cancellationToken)
+        {
+            return await ValidateAdd(request)
+                .Tap(AddPolicy);
```
Comment (kirillta) [inline Api/Services/Markups/SupplierMarkupPolicyManager.cs:47]: Shouldn't be a `Bind` here?

File: Api/Services/Markups/SupplierMarkupPolicyManager.cs [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+
+
+            Result ValidateAdd(SupplierMarkupRequest request)
+                => GenericValidator<SupplierMarkupRequest>.Validate(v =>
+                    {
+                        var valueValidatorMessage = "Markup policy value must be in range (-100..-0.1) or (0.1..100)";
```
Comment (kirillta) [inline Api/Services/Markups/SupplierMarkupPolicyManager.cs:53]: The class is quite big. Could you move validators to a separate entity?

File: Api/Services/Markups/SupplierMarkupPolicyManager.cs [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+        }
+
+
+        public async Task<Result> Modify(int policyId, SupplierMarkupRequest request, CancellationToken cancellationToken)
+        {
+            var policy = await _context.MarkupPolicies
+                .FirstOrDefaultAsync(p => p.Id == policyId, cancellationToken);
+
+            return await ValidateModify((request, policy))
+                .Tap(ModifyPolicy);
```
Comment (kirillta) [inline Api/Services/Markups/SupplierMarkupPolicyManager.cs:132]: And another `Bind` here

File: Api/Services/Markups/Validators/SupplierMarkupValidators.cs [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace Api.Services.Markups.Validators
+{
+    public static class SupplierMarkupValidators
+    {
+        public static Result ValidateAdd(SupplierMarkupRequest request, EdoContext context)
+                => GenericValidator<SupplierMarkupRequest>.Validate(v =>
+                    {
+                        var valueValidatorMessage = "Markup policy value must be in range (-100..-0.1) or (0.1..100)";
```
Comment (kirillta) [inline Api/Services/Markups/Validators/SupplierMarkupValidators.cs:20]: Mat be a const

File: Api/Services/Markups/Validators/SupplierMarkupValidators.cs [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+                            .When(r => r.DestinationScopeType == DestinationMarkupScopeTypes.Country, ApplyConditionTo.CurrentValidator)
+                            .WithMessage(r => $"Destination markup policy with DestinationScopeId {r.DestinationScopeId} already exists or unexpected value!")
+                            .MustAsync(CountryExists()!)
+                            .When(m => m.DestinationScopeType == DestinationMarkupScopeTypes.Country, ApplyConditionTo.CurrentValidator)
+                            .WithMessage(m => $"Country with code {m.DestinationScopeId} doesn't exist!");
+
+                        v.RuleFor(m => m.DestinationScopeType)
+                            .NotNull()
+                            .Must(d => d.Equals(DestinationMarkupScopeTypes.Country))
+                            .WithMessage($"Request's destinationScopeType must be Country");
```
Comment (kirillta) [inline Api/Services/Markups/Validators/SupplierMarkupValidators.cs:44]: The string isn't interpolated. `$` isn't needed here and below

**Pull request merged**